### ISSUE TITLE
problem with point-of-instantiation

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#6252]
Update 2020-06: this issue is now about the confusing semantics. Originally, in 2017, it was opened as an "incorrect implementation" issue.

The relevant spec rule is defined [here](https://chapel-lang.org/docs/language/spec/generics.html#function-visibility-in-generic-functions). Note that it has a special definition for methods that can be invoked through dynamic dispatch.

See also #8077 - it presents function hijacking, which arises due to the POI rule.

Chapel currently tries to use a point-of-instantiation rule for generic instantiation, so that any functions available at the call site are available to the instantiated generic. However, this approach is not implemented correctly and it is not clear if point-of-instantiation is the right rule to have here.

Consider the program below, which declares a generic function, `callFoo`, which in turn is instantiated in two different ways to call different `foo` functions. 

``` chapel
module MyGeneric {
  // uses foo(param)
  proc callFoo(param x) {
    foo(x);
  }
}

module A {
  use MyGeneric;
  proc foo(param x) {
    writeln("In A.foo(", x, ")");
  }
  proc testA() {
    callFoo(1);
  }
}

module B {
  use MyGeneric;
  proc foo(param x) {
    writeln("In B.foo(", x, ")");
  }
  proc testB() {
    callFoo(1);
  }
}

module Main {
  proc main() {
    A.testA();
    B.testB();
  }
}
```

This program prints out
```
In A.foo(1)
In A.foo(1)
```

but one would expect it to output
```
In A.foo(1)
In B.foo(1)
```

One issue here is that generic instantiation in Chapel uses a cache, so that generic functions are only instantiated once for a given set of types and params. However, this cache is not aware of scoping. That means that in this case, the version instantiated first is re-used for the second call, even though it should be re-instantiated if we completely use a point-of-instantiation rule.

It is not clear to me if we want to keep point-of-instantiation at all, and if so, if we need it to work for this example. If we do keep it, adding some knowledge of scopes to the generics cache is one potential path.

**Comment by lydia-duncan** [comment]
Yeah, I'm not surprised by this, sadly.  I do think that the program you've described should act as though it has been instantiated in both locations instead of just one.

**Comment by mppf** [comment]
@lydia-duncan - but it also might be reasonable to require one use an explicit mechanism if this is desired. I.e., in most cases, isn't point-of-definition OK? Then in other cases, like Sort, we could use  interface/implements to explicitly indicate that the required function should be provided by the calling context.

**Comment by lydia-duncan** [comment]
It just occurred to me, maybe we have this unusual handling so that reuse of the instantiation is even possible?  As in, if we placed the instantiation at the callsite, that might be too limited a scope for it to be useful anywhere else, even if it would be valid to reuse that instantiation elsewhere.

Say we're in the program you've written above.  If callFoo can be instantiated at all when it is called, we would have to place that instantiation either within the body of the functions which called it (in which case, for this program no one can reuse it since it won't be visible outside that scope) or at the definition site.  In the former case, a module that `use`s A and tries to use the same instantiation of callFoo would have to recreate it locally, potentially adding O(number of calls) copies of this instantiation in the worst case scenario.  If we try to be smart and place the instantiation we've created at a scope where it is accessible to others but local to the call, how do we determine where that location should?  Based on the defining scope for the closest symbol it has resolved to use?  That'd be a lot of information to keep around for resolving every generic function.

**Comment by mppf** [comment]
@lydia-duncan - I'm sure that correct point-of-instantiation would be more complex to implement and/or result in less sharing of instantiations. But I am trying to ask - what do we want the language to be?

**Comment by lydia-duncan** [comment]
The trouble is, I'm not entirely certain.  I don't like the idea of definitions at the call site taking precedence over definitions local to the function definition in general, which is what happens today, but I do think it is useful to do so when there's a type defined local to the call site which the called function is not otherwise aware of.  I think there's a lot of edge cases to this problem that aren't even occurring to me, too.

**Comment by mppf** [comment]
@lydia-duncan - I don't think that this is an implementation issue; rather it is a case where the implementation is behaving according to the specification. Given that, I think I will close the issue.

Spec section 22.2:

> When resolving function calls made within generic functions, the visible functions are taken from any call site at which the generic function is instan- tiated for each particular instantiation. The specific call site chosen is arbitrary and it is referred to as the point of instantiation.


Also, regarding this:

> I don't like the idea of definitions at the call site taking precedence over definitions local to the function definition in general, which is what happens today

I don't think that's accurate. Consider this example:
``` chapel
module M1 {
  proc foo() {
    writeln("M1.foo");
  }
  proc generic(param p) {
    foo();
  }
}

module M2 {
  // results in ambiguity unless private
  proc foo() {
    writeln("M2.foo");
  }
}

module Test {
  use M1, M2;

  proc main() {
    // Is this function going to find M1's foo or M2's foo?
    //  ambiguity error -> prefer neither.
    //  M1.foo -> prefer point of definition
    //  M2.foo -> prefer point of instantiation
    generic(1);
  }
}
```

it results in a compilation failure due to ambiguity - so neither the call site nor the definition point are preferred. Other examples are just showing that a match with closer types is preferred over a more closely scoped ("more visible") match.

**Comment by lydia-duncan** [comment]
But the function foo is concrete in that example, so isn't subject to the point-of-instantiation rules.  What happens in the case where the function has more concrete arguments, but still some generic ones?

**Comment by mppf** [comment]
Same thing. Chapel's rules don't depend on whether the called function is generic or not (but C++ rules *do*).

``` chapel
module M1 {
  proc foo(param p) {
    writeln("M1.foo", p);
  }
  proc generic(param p) {
    foo(p);
  }
}

module M2 {
  // results in ambiguity unless private
  proc foo(param p) {
    writeln("M2.foo", p);
  }
}

module Test {
  use M1, M2;

  proc main() {
    // Is this function going to find M1's foo or M2's foo?
    //  ambiguity error -> prefer neither.
    //  M1.foo -> prefer point of definition
    //  M2.foo -> prefer point of instantiation
    generic(1);
  }
}
```

```
$ chpl ex4.chpl 
ex4.chpl:5: In function 'generic':
ex4.chpl:6: error: ambiguous call 'foo(1)'
ex4.chpl:6: note:   instantiated from generic
ex4.chpl:2: note: candidates are: foo(param p: int(64)) [1025816]
ex4.chpl:12: note:                 foo(param p: int(64)) [1025823]
```

**Comment by lydia-duncan** [comment]
I think I'm confused about your position.  Are you saying that this issue is not accurate because it is the specification that needs to be changed rather than the implementation?  Are you saying that we've misinterpreted the implementation and that it is closer to the behavior we actually want than we thought?  Or something else entirely?

**Comment by mppf** [comment]
I think the specified behavior is OK, and since it's specified that way, the program this issue started with works in "a reasonable manner".

aka
>we've misinterpreted the implementation and that it is closer to the behavior we actually want than we thought

**Comment by lydia-duncan** [comment]
Okay.  I'm not convinced, though, because in the cases you are demonstrating, it seems likely that the writer of M1 wanted M1.foo to be called all the time and giving an ambiguity error or finding M2.foo is breaking their assumptions about their function in a subtle way that won't be apparent until a potentially malicious client is using their module.

So I think the specification is wrong and can lead to confusion for users.  I think we still want "point-of-definition except if no match is found when resolving from the definition point of the function, in which case look from the caller's scope".

**Comment by lydia-duncan** [comment]
Yes, their assumptions in the current code would be based on false premises because it would be contrary to the specification, but the different between a code that would work the way they expect and one that doesn't is very subtle and error-prone, so I feel there's more we should be doing to avoid putting the user in that position unnecessarily and don't see very much benefit from the other approach

**Comment by mppf** [comment]
See also https://dlang.org/hijack.html

**Comment by lydia-duncan** [comment]
Thanks for the link, so far (I'm about halfway through), this looks closer to what I would like to see than the point-of-instantiation rules we have today.

**Comment by mppf** [comment]
I realized that my earlier comment on this thread was wrong. The compiler / language *does* prefer the point of instantiation to the point of definition.
Here's a corrected example:
``` chapel
module M1 {
  proc foo() {
    writeln("M1.foo");
  }
  proc generic(param p) {
    foo();
  }
}

module Main {
  use M1;

  proc foo() {
    writeln("Main.foo");
  }

  proc main() {
    // Is this function going to find M1's foo or M2's foo?
    //  ambiguity error -> prefer neither.
    //  M1.foo -> prefer point of definition
    //  Main.foo -> prefer point of instantiation
    generic(1);
  }
}
```

**Comment by vasslitvinov** [comment]
Reopening to discuss the confusing semantic itself.

**Comment by vasslitvinov** [comment]
Added user-code scenarios in [a gist](https://gist.github.com/vasslitvinov/148ff3abf018b693ae738b8fa2e09ba9) .

**Comment by vasslitvinov** [comment]
Here is a survey of realistic code scenarios in our Chapel code base that rely on POI.

<details open><summary>click to minimize/maximize</summary>

#### Test directories of note

Here are the more prominent tests / test directories:
```
test/functions/generic/poi/
test/functions/ferguson/hijacking/
test/visibility/private/uses/pointOfInstantiation/
test/npb/cg/bradc/cg-sparse-partred.chpl
test/release/examples/spec/Generics/point-of-instantiation.chpl
```

#### Basic scenarios

The POI rule was introduced to allow a generic function to invoke operations defined in the caller scope. The following examples "do just that."

`test/functions/generic/poi/hashed-dist-with-mapper.chpl` : the implementation of HashedDist invokes a method on the user-defined mapper to determine the locale for the particular index upon a call to domain.add(). The mapper is stored in the domain descriptor and is not passed directly to the add() call. The method in question is visible from the scope of the add() caller, not from from the HashedDist implementation.

`test/functions/generic/poi/init-in-private-array.chpl` : the implementation of PrivateDist creates a new array using default initialization i.e. init() on its elements. If the array elements are of a user-defined record type, init() is defined in the user space and is not visible to PrivateDist implementation.

`test/functions/generic/poi/writethis-user-record.chpl` : the implementation of writeln() invokes writeThis() on the arguments. If an argument is a user-defined record, its writeThis() is defined in the user space, even if it is compiler-generated. In other cases, `<~>` is needed.

`test/functions/generic/poi/reduce-user-record.chpl` : reduce implementation calls functions defined in user space when reduced elements are user-defined records.

`test/functions/generic/poi/zmq-initequals.chpl` : ZMQ.recv invokes init=() or `=` on a user-defined record type.

#### Other scenarios

**canResolve()** et al. is most likely to be affected by the issue in [#6252](https://github.com/chapel-lang/chapel/issues/6252) because of how it is currently implemented. We may want to implement it differently, ex. similarly to compilerError(). Test: `test/functions/generic/poi/canresolve-in-sort.chpl`

`test/npb/cg/bradc/cg-sparse-partred.chpl` have a reduction implementation invoke a this() method defined in the application in a local scope. It is notable because the method is defined in a local scope. Which is essential to enable the method to access the local variables in that scope. A small reproducer is in `test/functions/generic/poi/reduce-nested-method.chpl`

`test/functions/generic/poi/forwarding-wrapper.chpl` summarizes the reliance of our array implementation on POI in conjunction with forwarding, observed in [#10358](https://github.com/chapel-lang/chapel/issues/10358)

`test/functions/ferguson/hijacking/ApplicationB.chpl` is the example in #6252. It is also an example of application development, where an earlier version could have been the one in `ApplicationA.chpl`.

See [#15714](https://github.com/chapel-lang/chapel/issues/15714) .

In many cases, the reliance on POI can be avoided by importing the target functions explicitly. Examples:

* variants of cholesky in `test/users/jglewis`

* `test/distributions/bharshbarg/stencil/periodic.chpl` - it uses its helper module 'util', as well as StencilDist; the 'util' module accesses `fluff` in StencilDist without importing it

* quickSort() in the standard module `Sort` - it is illustrated in `test/functions/generic/poi/sort-check-comparator.chpl`; a similar scenario is in `test/functions/generic/poi/random-choice.chpl`

* the internal modules `MemConsistency` and `AlignedTSupport` (in ChapelSyncvar.chpl) - they invoke _cond_test() and _cond_invalid() without importing them from ChapelBase

</details>

**Comment by vasslitvinov** [comment]
How can  the scenarios from the previous comment be converted to Constrained Generics (CG)? What can we learn from them about POI?

<details open><summary>click to minimize/maximize</summary>

#### test/functions/generic/poi/canresolve-in-sort.chpl

`canResolve` and similar Reflection functions are more likely to be affected by [#6252](https://github.com/chapel-lang/chapel/issues/6252). However, that is an artifact of how they are implemented. A different implementation - ex. similar to compilerError() - can remove reliance of `canResolve` on POI. Which would have no user-observable impact, except for restored correctness in corner cases.

#### test/functions/generic/poi/forwarding-wrapper.chpl

Similar scenario: **test/functions/generic/poi/init-in-private-array.chpl** and **test/functions/generic/poi/zmq-initequals.chpl**  The test `forwarding-wrapper.chpl` adds forwarding to the call chain that POI needs to consult. The following discusses the scenario without forwarding.

Consider this simple example from [#10358](https://github.com/chapel-lang/chapel/issues/10358) :

```chpl
record R {}
var bag = new DistBag(R);  // in DistributedBag
bag.add(new R());
```

<details><summary>a highlight of DistributedBag</summary>

```chpl
module DistributedBag {
  record DistBag {
    type eltType;
    var seg = new BagSegmentBlock(eltType); // a simplification
    proc add(elt: eltType) {
      this.seg.push(elt);
    }
  }
  class BagSegmentBlock {
    type eltType;
    var elems: eltType; // a simplification
    proc push(elt: eltType) {
      elems = elt;      // invokes R.init= or assignment
    }
  }
}
```
</details>

The library implementation of `DistBag.add()` needs assignment between `R` records. Collection libraries in general may also need other predefined methods on their elements, such as `init`, `init=`, and `deinit`. These methods for a user-defined record will be defined in the application code and so will not be immediately visible to the library.

The design question here is: do we treat these special methods the same as any other user-defined function, or do we do something different?

For example, every record must have a `deinit`, and in most cases it is easily discoverable by the compiler from the record type. It looks superfluous to have the library writer discuss it in a constraint.

The initializers and the assignment are less clear-cut, as some of them can be missing or overloaded for different argument types. For example, does the collection require a default initializer? If no, is a call to a default initializer within the collection implementation an error? One option is for the user to specify this in a constraint; another is for the compiler to infer it -- which is closer to how unconstrained generics work.

#### test/functions/generic/poi/writethis-user-record.chpl

In a sentence, the implementation of`writeThis` or similar for a user record needs to be accessed by the library implementation of `writeln()`.

The twist here is that `writeThis` is available for **every** type. So a CG constraint would not actually restrict any call to writeln() from getting compiled.

Still, if the IO library is switched to CG, the compiler will need the "witness" for `writeThis` because otherwise it will not be visible to the `writeln` implementation in the case of a user-defined type. So we may as well add a "writeThis is provided" constraint where needed in the IO library.

#### test/functions/generic/poi/hashed-dist-with-mapper.chpl

Converting it to CG is straightforward. For example:
```chpl
// HashedDist module introduces an interface for the mapper object
interface HashedMappable(type Self) {
  proc Self.this(elm /*what is the type of 'elm'?*/, targetLocs: [] locale);
}

// This interface is required when creating a 'Hashed'
proc Hashed.init(..., mapper: HashedMappable) {
  ......
}

// The compiler infers that the mapper argument to Hashed.init
// implements this interface. The application code does not change.
var myMapper = .....;
var D: domain(string) dmapped Hashed(idxType=string,
                                     mapper=myMapper);
```

Here is an interesting piece. The type `Hashed` will need to know that its field `mapper` implements the interface `HashedMappable` in the above example. Perhaps that can be inferred from `Hashed.init()` ? Such inference might be complicated by the presence of a copy-like initializer - used for privatization - where ideally the library writer would not need to indicate this constraint for the source of the copy.

From `Hashed.mapper`, the "implements HashedMappable" knowledge probably needs to be transferred to fields of other classes that `Hashed` creates, such as `UserMapAssocDom.mapperType`. Even if such transfer is not needed in this particular case, there are other cases where it IS needed, for example in the module `DistributedBag` from `DistBag.eltType` through to `BagSegmentBlock.eltType`.

#### test/functions/generic/poi/reduce-user-record.chpl

The (internal) implementation of product reduction needs to impose a requirement on `ProductReduceScanOp.eltType` to implement the following interface:
```chpl
interface Multipliable(type Self) {
  proc identity(type Self): Self;
  proc *(a: Self, b: Self): Self;
}
```
Cf. the current implementation actually requires a cast from `1` to `eltType` (see `_prod_id()` in Types.chpl) and `*=(eltType,eltType)`. We will probably prefer something closer to the above and so adjust the code in `ProductReduceScanOp` in trival ways.

Again, the application code needs not change, as long as the compiler infers interface compliance at call site.

</details>

**Comment by vasslitvinov** [comment]
The semantics is defined in #15872 : function lookup is done as if POI (the Point of Instantiation) is taken specifically for each Chapel call chain.

The implementation approach is proposed in #15923, relying on a simplification proposed in #15948 and implemented in #16279.

This semantics and the implementation approach are further explained and implemented in #16261 and #16450 and #16492.

Closing this issue in favor of specific remaining issues, currently #16502.
